pandas/core/frame.py

```diff
@@ -6482,7 +6482,22 @@ def combiner(x, y):
 
             return expressions.where(mask, y_values, x_values)
 
-        return self.combine(other, combiner, overwrite=False)
+        combined = self.combine(other, combiner, overwrite=False)
+
+        dtypes = {
+            col: find_common_type([self.dtypes[col], other.dtypes[col]])
+            for col in self.columns.intersection(other.columns)
+            if not is_dtype_equal(combined.dtypes[col], self.dtypes[col])
+            and not is_dtype_equal(
```

Comment:
why do u have this complicated condition?

Reply:
first if check for columns that are candidates for preserving the dtype, and the second if avoids extra computation if the common dtype is the same as the combined dtype

Reply:
ok no need for that, as astype will handle this if you pass `copy=False`, though i actually don't think this really matters, i would just remove the redudant check

Reply:
done